src/HelperTraits/SyncTrait.php

```diff
@@ -0,0 +1,135 @@
+<?php
+declare( strict_types=1 );
+
+namespace Automattic\WooCommerce\GoogleListingsAndAds\HelperTraits;
+
+use Automattic\WooCommerce\GoogleListingsAndAds\Jobs\MigrateGTIN;
+use Automattic\WooCommerce\GoogleListingsAndAds\Options\OptionsAwareTrait;
+use Automattic\WooCommerce\GoogleListingsAndAds\Options\OptionsInterface;
+use Exception;
+use WC_Product;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Trait with some helpers for API Pull and MC Push Sync modes.
+ *
+ * @package Automattic\WooCommerce\GoogleListingsAndAds\HelperTraits
+ */
+trait SyncTrait {
+
+	use OptionsAwareTrait;
+
+	/**
+	 * Get the products datatype key
+	 *
+	 * @return string
+	 */
+	protected function get_products_datatype(): string {
+		return 'products';
+	}
+
+	/**
+	 * Get the coupons datatype key
+	 *
+	 * @return string
+	 */
+	protected function get_coupons_datatype(): string {
+		return 'coupons';
+	}
+
+	/**
+	 * Get the shipping datatype key
+	 *
+	 * @return string
+	 */
+	protected function get_shipping_datatype(): string {
+		return 'shipping';
+	}
+
+	/**
+	 * Get the settings datatype key
+	 *
+	 * @return string
+	 */
+	protected function get_settings_datatype(): string {
+		return 'settings';
+	}
+
+	/**
+	 * Get the default config for the sync mode.
+	 *
+	 * @return array[]
+	 */
+	protected function get_default_sync_mode(): array {
+		return [
+			$this->get_products_datatype() => [
+				'pull' => true,
+				'push' => true,
+			],
+			$this->get_coupons_datatype()  => [
+				'pull' => true,
+				'push' => true,
+			],
+			$this->get_shipping_datatype() => [
+				'pull' => true,
+				'push' => true,
+			],
+			$this->get_settings_datatype() => [
+				'pull' => true,
+				'push' => true,
+			],
+		];
+	}
+
+	/**
+	 * Get the current value for the API PULL Sync.
+	 * Notice that malformed data will be replaced by default data.
+	 *
+	 * @return array
+	 */
+	protected function get_current_sync_value(): array {
+		$sync_mode = $this->options->get( OptionsInterface::API_PULL_SYNC_MODE );
+
+		if ( ! is_array( $sync_mode ) ) {
+			$sync_mode = $this->get_default_sync_mode();
+		}
+
+		return array_replace_recursive( $this->get_default_sync_mode(), $sync_mode );
+	}
+
+	/**
+	 * Check if API PULL is enabled for a specific data type.
+	 *
+	 * Checking null data type returns true.
+	 * Checking a non-existent data type will return false.
+	 *
+	 * @param string|null $data_type The data type to check.
+	 * @return bool
+	 */
+	protected function is_pull_enabled_for_datatype( string $data_type = null ): bool {
+		if ( is_null( $data_type ) ) {
+			return true;
+		}
+
+		$sync_modes = $this->get_current_sync_value();
+
+		return apply_filters( 'woocommerce_gla_is_pull_enabled_for_datatype', $sync_modes[ $data_type ]['pull'] ?? false );
```

Comment:
Always best to cast to expected type after applying a filter as we don't know what type a filter might return.

```suggestion
		return (bool) apply_filters( 'woocommerce_gla_is_pull_enabled_for_datatype', $sync_modes[ $data_type ]['pull'] ?? false );
```

Curious how useful this filter is used. If someone actually intended to use this then wouldn't they need the `$datatype` passed to the filter to check if they want to override it? Also wouldn't it be handier for them to have a filter to the function `get_current_sync_value` if they needed to override sync values?

If the prime reason to add this filter is just for the unit testing, then I'd suggest we don't add the filter. In this case we can mock the value returned from `$this->options` to mimic a specific mode for a particular datatype.

Same applies for the `is_push_enabled_for_datatype` function.

Reply:
Thanks @mikkamp 

>  If someone actually intended to use this then wouldn't they need the $datatype

Yes, correct, I missed the datatype

> If the prime reason to add this filter is just for the unit testing

The reason is to be able enable/disable push/pull using the filter for a specific datatype or all of them.

> So wouldn't it be handier for them to have a filter to the function get_current_sync_value if they needed to override sync values?

In practice this filter does the same. The difference is that you can enable or disable the method directly instead of overriding the values.

Reply:
I added the new filter anyway.

Reply:
https://github.com/woocommerce/google-listings-and-ads/pull/2804/commits/19fad8762198794837af6c2672a4e83e7dfc95cf and